chain/network/src/accounts_data/mod.rs

```diff
@@ -156,45 +134,38 @@ impl Cache {
         // Filter out non-interesting data, so that we never check signatures for valid non-interesting data.
         // Bad peers may force us to check signatures for fake data anyway, but we will ban them after first invalid signature.
         // It locks epochs for reading for a short period.
-        let mut found = HashSet::new();
-        let mut data_and_keys = vec![];
-        let epochs = self.epochs.read();
+        let mut data_and_keys = HashMap::new();
+        let inner = self.inner.read();
         for d in data {
-            // We want the communication needed for broadcasting per-account data to be minimal.
-            // Therefore broadcasting multiple datasets per account is considered malicious
-            // behavior, since all but one are obviously outdated.
-            let data_id = (d.epoch_id.clone(), d.account_id.clone());
-            if found.contains(&data_id) {
-                return (vec![], Some(Error::SingleAccountMultipleData));
-            }
-            found.insert(data_id);
             // There is a limit on the amount of RAM occupied by per-account datasets.
             // Broadcasting larger datasets is considered malicious behavior.
             if d.payload().len() > network_protocol::MAX_ACCOUNT_DATA_SIZE_BYTES {
                 return (vec![], Some(Error::DataTooLarge));
             }
-            // It is fine to broadcast account data for epochs that we don't care about.
-            let epoch = match epochs.0.get(&d.epoch_id) {
-                Some(e) => e,
-                None => continue,
-            };
-            // It is NOT fine to broadcast account data which is not relevant to the given epoch.
-            let account = match epoch.0.get(&d.account_id) {
-                Some(a) => a,
-                None => return (vec![], Some(Error::InvalidAccount)),
-            };
+            let id = (d.epoch_id.clone(), d.account_id.clone());
+            // We want the communication needed for broadcasting per-account data to be minimal.
+            // Therefore broadcasting multiple datasets per account is considered malicious
+            // behavior, since all but one are obviously outdated.
+            if data_and_keys.contains_key(&id) {
+                return (vec![], Some(Error::SingleAccountMultipleData));
+            }
             // It is fine to broadcast data we already know about.
-            if !account.is_new(d.timestamp) {
+            if !inner.is_new(&d) {
                 continue;
             }
-            data_and_keys.push((d, account.key.clone()));
+            // It is fine to broadcast account data that we don't care about.
+            let key = match inner.keys.get(&id) {
+                Some(key) => key.clone(),
```

Comment:
Yes, there is a reason. Current implementation removes one level of indirection (set of epochs) and removes all the semantic assumptions about the epoch from the network: now EpochId is just an identifier, and we do care (on the network side) that the EpochId uniquely determines the set of account keys. It makes sense because network is a dependency of the client module , not the other way around (network module provides a transportation layer and it should be as chain logic agnostic as possible).